Ramos_HwAssignments/Ramos_ClassAssignment_BarcodeScanner/BarcodeScanner.tsx

```diff
@@ -0,0 +1,71 @@
+import React, { useState, useEffect } from 'react';
+import { Text, View, StyleSheet, Button } from 'react-native';
+import { BarCodeScanner } from 'expo-barcode-scanner';
+import { NavigationContainer, useNavigation, useRoute, RouteProp, StackActions } from '@react-navigation/native';
+import { StackNavigationProp } from '@react-navigation/stack';
+import { StackParamList } from './App';
+import { useFetch } from './Functions/useFetch';
+
+type ScannerNavigationProp = StackNavigationProp<ScannerStackParamList, 'BarcodeScanner'>;
+
+export const BarcodeScanner = () => {
+  const navigation = useNavigation<ScannerNavigationProp>();
+
+  const [hasPermission, setHasPermission] = useState(null);
+  const [scanned, setScanned] = useState(false);
+
+  const defaultURL = {data: "https://rn-products-1d8a6-default-rtdb.firebaseio.com/products/1.json"};
+  const [productURL, setProductURL] = useState("");
+  const {data, loading, error} = useFetch(productURL);
+
+  useEffect(() => {
+    const getBarCodeScannerPermissions = async () => {
+      const { status } = await BarCodeScanner.requestPermissionsAsync();
+      setHasPermission(status === 'granted');
+    };
+
+    getBarCodeScannerPermissions();
+  }, []);
+
+  useEffect(() => {
+    if (data != null)
+      navigation.navigate('ProductDetail', {
```

Comment:
Thanks for sharing your line of thinking about the barcode scanner fetching the data instead of product details. I enjoy having these types of conversations. I want to share my opinion here too. Mainly I want to focus here:

> The line of thinking I'm on is separation of concerns. The product details page shouldn't be concerned with the fetching qr urls.

In my opinion the bar code scanner is now doing too much. The scanner obviously scans a bar code and gets some data, but then also fetches additional data from an API? To me, this is where we should separate our concerns.

The job of the barcode scanner is to scan a barcode (like I needed to tell you that 😝 ). It can then pass that information (a URL or even a product ID) to a screen that has a different job. In the case of `ProductDetail`, it's one job is to fetch that API data and it can also check if your item has been favorited or not

Ideally, you'd want to pass the minimal amount of information to a screen to get it to work. Think about routing in web development. If we wanted to see a list of products we'd visit `www.mysite.com/products/` and if we wanted to visit a single product (a product detail) we'd visit `www.mysite.com/products/1`. We can visit `www.mysite.com/products/1` directly because that screen fetches its own data. Although this is not a exact 1:1 example, using the current structure on your app means that we'd have to visit `www.mysite.com/products` first before reaching `www.mysite.com/products/1`.

There's something called deep linking that we don't cover in this course, but it's a way to link to specific screens in your app. This is how tapping a web url can redirect you to a screen in your app. If the `ProductDetail` screen depends on the previous screen to scan a barcode first, the we can never directly visit the screen.

Hopefully that all made sense.